packages/backend-common/src/service/lib/hostFactory.ts

```diff
@@ -95,11 +95,21 @@ async function getGeneratedCertificate(hostname: string, logger?: Logger) {
   }
 
   let cert = undefined;
+  let remainingMs = 0;
   if (await fs.pathExists(certPath)) {
-    const stat = await fs.stat(certPath);
-    const ageMs = Date.now() - stat.ctimeMs;
-    if (stat.isFile() && ageMs < ALMOST_MONTH_IN_MS) {
-      cert = await fs.readFile(certPath);
+    cert = await fs.readFile(certPath);
+    try {
+      const forge = require('node-forge');
+      const crt = forge.pki.certificateFromPem(cert);
+      const crtTimestamp = Date.parse(crt.validity.notAfter);
+      remainingMs = crtTimestamp - Date.now();
+    } catch (error) {
+      logger?.warn(`Unable to parse self-signed certificate. ${error}`);
+      remainingMs = 0;
```

Comment:
I had different implementations of this error-catching block and the comparison in line 110 afterwards.
I came up with this, because I think that the condition in line 110 with `remainingMs < FIVE_DAYS_IN_MS` is now easy to understand.
So, I also wanted to be sure that it's easy to understand that errors will reset the `remainingMs` so that the comparison in line 110 will fail and a new certificate will be generated.

Are you suggesting to move the "complexity" into the comparison of line 110?

Reply:
Sorry if I was unclear, the certificate parsing logic could be extracted into a function to make readability a bit better.

I'm thinking a function like `getCertificateExpiration(cert: string): Date`

```ts

const remaining = getCertificateExpiration(cert) - Date.now()
if (remainingMs < FIVE_DAYS_IN_MS) {
  cert = undefined;
}
```

Reply:
Thanks, understood. 
Will look into it.

Reply:
@jhaals I refactored that part.
Unfortunately I can't judge why [Node 14.x on windows-2019](https://github.com/backstage/backstage/actions/runs/3308515761/jobs/5460953482#logs) is failing.

Reply:
I found this https://github.com/yarnpkg/berry/issues/4985
So, it seems to be a bug of `json-schema-library@7.2.0`.

Reply:
Rebased again and tests are fine now @jhaals